plugins/flashrom/fu-plugin-flashrom.c

```diff
@@ -184,19 +206,58 @@ fu_plugin_update (FuPlugin *plugin,
 	if (!fu_common_set_contents_bytes (firmware_fn, blob_fw, error))
 		return FALSE;
 
-	/* use flashrom to write image */
-	argv[4] = firmware_fn;
-	if (!fu_common_spawn_sync ((const gchar * const *) argv,
-				   fu_plugin_flashrom_write_cb, device,
-				   NULL, error)) {
-		g_prefix_error (error, "failed to write firmware: ");
+	if (flashrom_programmer_init(&flashprog, "internal", NULL)) {
+		g_error ("Error: Programmer initialization failed.");
+		return FALSE;
+	}
+
+	if (flashrom_flash_probe(&flashctx, flashprog, NULL)) {
+		g_error ("Error: Flash probe failed.");
+		return FALSE;
+	}
+
+	flashrom_flag_set(flashctx, FLASHROM_FLAG_VERIFY_AFTER_WRITE, TRUE);
+
+	if (flashrom_layout_read_from_ifd(&layout, flashctx, NULL, 0)) {
+		g_error ("Error: Reading layout from Intel ICH descriptor failed.");
+		return FALSE;
+	}
+
+	/* Include bios region for safety reasons */
+	if (flashrom_layout_include_region(layout, "bios")) {
+		g_error ("Error: Invalid region name.");
+		return FALSE;
+	}
+
+	flashrom_layout_set(flashctx, layout);
+
+	const size_t flash_size = flashrom_flash_getsize(flashctx);
+
+	uint8_t *newcontents = malloc(flash_size);
+	if (!newcontents) {
+		g_error ("Error: Out of memory.");
+		return FALSE;
+	}
+
+	if (read_buf_from_file(newcontents, flash_size, firmware_fn)) {
+		free(newcontents);
+		return FALSE;
+	}
+
+	if (flashrom_image_write(flashctx, newcontents, flash_size, NULL)) {
+		g_error ("Error: Image write failed.");
 		return FALSE;
 	}
 
 	/* delete temp location */
 	if (!fu_common_rmtree (tmpdir, error))
 		return FALSE;
 
+	g_free(newcontents);
+	flashrom_layout_release(layout);
+	flashrom_programmer_shutdown(flashprog);
+	flashrom_flash_release(flashctx);
```

Comment:
Drop the g_free (per above comment).

The others I think these should be moved into a device close routine rather than part of the write.  That way after you return TRUE or FALSE they always get ran.